---
title: "DADA2 tutorial pipeline"
output: html_document
---

## Set up
```{r setup, include=FALSE}
library(Rcpp)
library(dada2); packageVersion("dada2")
#library(phyloseq); packageVersion("phyloseq")
library(parallel)
library(BiocGenerics)
library(Biostrings); packageVersion("Biostrings")
library(ggplot2); packageVersion("ggplot2")
library(ShortRead); packageVersion("ShortRead")
```

## Import data
```{r}
path <- "~/Desktop/demux"
filenames = list.files(path)
```

```{r}
# Forward and reverse fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq
fnFs <- sort(list.files(path, pattern=".f.fastq", full.names = TRUE))
fnRs <- sort(list.files(path, pattern=".r.fastq", full.names = TRUE))
```

```{r}
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names1 <- sapply(strsplit(basename(fnFs), ".fastq"), `[`, 1)
sample.names2 <- sapply(strsplit(basename(fnRs), ".fastq"), `[`, 1)
```

```{r}
derep-class$quals
```


```{r}
## Inspect read quality profiles
plotQualityProfile(fnFs[41])
plotQualityProfile(fnRs[41])
```

## Filter and trim (1.16)
```{r}
# Place filtered files in filtered/ subdirectory
filtFs <- file.path(path, "filtered", paste0(sample.names1, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names2, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names1
names(filtRs) <- sample.names2
```

```{r}
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs,
maxN=0, maxEE=c(2,2), truncQ=2, trimLeft=15, rm.phix=TRUE,
compress=TRUE, multithread=TRUE)
head(out)
```

```{r}
## Learn the error rate
errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)
plotErrors(errF, nominalQ=TRUE)
```

```{r}
## Sample Inference
filtFs <- filtFs[file.exists(filtFs)]
filtRs <- filtRs[file.exists(filtRs)]
dadaFs <- dada(filtFs, err=errF, multithread=TRUE)
dadaRs <- dada(filtRs, err=errR, multithread=TRUE)

## Merge paired reads
merge <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)
```


```{r}
## Construct sequence table
seqtab <- makeSequenceTable(merge)
dim(seqtab)

# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))


## Remove chimeras
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)

sum(seqtab.nochim)/sum(seqtab)
```

## Track reads through the pipeline
```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(merge, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names2
head(track)
```

```{r}
## Assign taxonomy
taxa <- assignTaxonomy(seqtab.nochim, "~/Desktop/demux/silva_nr99_v138.1_train_set.fa.gz", multithread=TRUE)

taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)
```

## Construct the phylogenetic tree
```{r}
seqs <- getSequences(seqtab.nochim)
names(seqs) <- seqs # This propagates to the tip labels of the tree
mult <- msa(seqs, method="ClustalW", type="dna", order="input")
```

```{r}
library("phangorn")
phang.align <- as.phyDat(mult, type="DNA", names=getSequence(seqtab))
dm <- dist.ml(phang.align)
treeNJ <- NJ(dm) # Note, tip order != sequence order
fit = pml(treeNJ, data=phang.align)

## negative edges length changed to 0!

fitGTR <- update(fit, k=4, inv=0.2)
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
                       rearrangement = "stochastic", control = pml.control(trace = 0))
detach("package:phangorn", unload=TRUE)
```


## Evaluate accuracy
```{r}
unqs.mock <- seqtab.nochim["Mock",]
unqs.mock <- sort(unqs.mock[unqs.mock>0], decreasing=TRUE) # Drop ASVs absent in the Mock
cat("DADA2 inferred", length(unqs.mock), "sample sequences present in the Mock community.\n")

mock.ref <- getSequences(file.path(path, "HMP_MOCK.v35.fasta"))
match.ref <- sum(sapply(names(unqs.mock), function(x) any(grepl(x, mock.ref))))
cat("Of those,", sum(match.ref), "were exact matches to the expected reference sequences.\n")
```

## Bonus: Handoff to phyloseq
```{r}
theme_set(theme_bw())
samples.out <- rownames(seqtab.nochim)
subject <- sapply(strsplit(samples.out, "D"), `[`, 1)
gender <- substr(subject,1,1)
subject <- substr(subject,2,999)
day <- as.integer(sapply(strsplit(samples.out, "D"), `[`, 2))
samdf <- data.frame(Subject=subject, Gender=gender, Day=day)
samdf$When <- "Early"
samdf$When[samdf$Day>100] <- "Late"
rownames(samdf) <- samples.out
```

```{r}
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(samdf), 
               tax_table(taxa))
ps <- prune_samples(sample_names(ps) != "Mock", ps) # Remove mock sample
```

```{r}
dna <- Biostrings::DNAStringSet(taxa_names(ps))
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps)))
ps
```

# Visualize alpha-diversity:
```{r}
plot_richness(ps, x="Day", measures=c("Shannon", "Simpson"), color="When")
```

# Ordinate
```{r}
# Transform data to proportions as appropriate for Bray-Curtis distances
ps.prop <- transform_sample_counts(ps, function(otu) otu/sum(otu))
ord.nmds.bray <- ordinate(ps.prop, method="NMDS", distance="bray")

plot_ordination(ps.prop, ord.nmds.bray, color="When", title="Bray NMDS")
```

# Bar plot
```{r}
top20 <- names(sort(taxa_sums(ps), decreasing=TRUE))[1:20]
ps.top20 <- transform_sample_counts(ps, function(OTU) OTU/sum(OTU))
ps.top20 <- prune_taxa(top20, ps.top20)
plot_bar(ps.top20, x="Day", fill="Family") + facet_wrap(~When, scales="free_x")
```

```{r}
origin = data.frame(table(taxa_origin$Class))
new = data.frame(table(taxa_new$Class))
colnames(origin) = c('Class','Count')
colnames(new) = c('Class','Count')
compare = rbind(origin%>%mutate(type='origin'),new%>%mutate(type='rc'))
```

```{r}
compare%>%
ggplot()+
  geom_col(aes(x = Class,y=Count, fill = type), color = 'black')+
  theme(legend.position = 'none')+
  scale_y_sqrt()+
  theme(axis.text.x = element_text(angle = 90))

```
